### DISCUSSION TITLE
Implement Standard Schema interface

### DISCUSSION THREAD

**Discussion by marcomuser** [sinclairzx81/typebox#1152] [General]
Have you already considered implementing the standard schema interface: https://github.com/standard-schema/standard-schema. This would provide third-party libraries a uniform integration to automatically support multiple schema libraries at once. Currently, zod, valibot and arktype implement the interface. I'm working on a library integrating with schema libs and I would love to also integrate with typebox without writing a custom adapter.

**Comment by sinclairzx81** [comment]
@marcomuser Hi, thanks for the suggestion.

There is a provisional Adapter to Standard Schema being developed and documented at the following URL

https://github.com/sinclairzx81/typebox/tree/master/example/standard

I have a lot to say about Standard Schema, will outline my thoughts on it below.

---

### Standard Schema (Review)

Unfortunately, I won't be able to implement the Standard Schema interface on TypeBox types because doing so would require TypeBox to include non-standard keyword vocabulary, such as  `~standard` on schematics, as well as require it to augment schematics with TypeBox specific validation functions (which wouldn't make sense for users who use TypeBox solely with Ajv)

Overall, I am not very keen on the design of Standard Schema, not only for the reasons mentioned above, but primarily because the specification would rule out any library that separates schematics from validation. This is a principled design decision that libraries make (i.e. adhering to separation of concerns) and unfortunately Standard Schema would require such libraries to violate these principles.

```typescript
const result = schema['~standard'].validate(value) // Incorrect: Because schemas should not have 
                                                   // functions and should not validate themselves. 

const result = validate(schema, value)             // Correct: Schemas are parameters to validators
```
While the former is possible in Zod, Valibot and Ark (which do carry validation around on their types), it's not feasible for TypeBox as it only builds Json Schema, and where those schematics are intended to be passed to any Json Schema compliant validator.

```typescript
const result = (new Ajv()).validate({ type: 'string' }, value) 
//                                  ^
//                                  |
//                                  Optional: Replace with Type.String()
```

These are the main concerns I have with the specification (Jan 2025)

---

### More Issues

Beyond the above concerns, I have several additional issues with the specification:

- The naming is misleading, as API interfaces have nothing to do with schematics (unless the intent is to conflate these terms).
- The return type for `validate` suggests a `parse` operation. Additionally, the returned value requires initialization for each validation call, hindering optimization. The specification should consider separating `parse` and `validate` into distinct API calls.
- Validation errors should be accessible via a separate API call rather than always being generated. Validation errors are diagnostics that incur a processing cost and may not be necessary in all environments. A system may prefer a boolean result instead of detailed diagnostics under the "minimal-disclosure" or "give no hints" principle.
- Validation errors currently expand all errors into potentially large objects, with no DoS mitigation mechanism for types that may generate large error sets. An error function that returns an iterator may address this issue (but do think a larger discussion should be had around error generation)
- The async return for validation does not make sense and should be removed.
- The specification makes no provisions for JIT (Just-In-Time) evaluation.

---

### Adapter

Needless to say, with all of the above, it isn't reasonable for TypeBox to break compatibility with a already widely adopted industry standard specification (Json Schema) just to unify with a few popular JavaScript validation libraries. 

Irrespective of the ergonomics or perceived DX wins integrators are expected to get (which I expect was largely used to inform the specification design), I really don't consider Standard Schema to be forward looking enough in terms of overall design, or representative of the myriad of other ways integrators can approach uniform type library adoption in frameworks.

As such, an Adapter will be required.

```typescript
import { StandardSchema } from './standard' // is it really standard?

const T = StandardSchema(Type.Object({  // Breaks Json Schema compliance by augmenting
  x: Type.Number(),                     // a TypeBox schematic with functions and keywords
  y: Type.Number(),                     // required by the Standard Schema interface.
  z: Type.Number(),
}))
```

---

### Alternatives

If you are looking for an alternative to Standard Schema, I would recommend the following projects (which are supportive of a wider range of type libraries and validators)

Current Gold Standard

https://github.com/decs/typeschema

Something I am working on which is a generalization of Fastify Type Providers

https://github.com/sinclairzx81/type-adapters

Something else I am exploring

https://github.com/sinclairzx81/typebox-adapter

---

Hope this brings some insight into where TypeBox is at with regards to Standard Schema. Happy to discuss more on this thread if you have any questions though. 

I have been asked to comment on the Standard Schema repository. Unfortunately, it is very difficult for me to communicate aspects of the specification without me suggesting that the authors adopt a design that runs contrary to how their libraries are written. I do note that the specification is written in service to frameworks such as tRPC where the specification does a better job of integration than previously, but falls short of servicing frameworks that are built upon industry specifications like Json Schema.

There is a lot to discuss, but too broader context in which to discuss it. But hopefully the above communicates my thoughts around it.

Cheers
S

**Reply by mmkal** [reply]
Thanks for the detailed answer! I found this because I want to add typebox support to my library which currently supports only standard-schema (since it's trpc/orpc based): https://npmjs.com/package/trpc-cli

A few questions for you, if you're up for reopening this discussion:

The main one being: would you consider shipping the adapter you linked to here as part of typebox? The maintenance burden seems low, given standard-schema have explicitly documented that you can freely copy-paste the spec in the form of a typescript interface and that will work for StandardSchemaV1 forever. Whereas if I have to copy-paste the adapter into trpc-cli, I need to carefully make sure I specify my typebox peerDependencies range properly, rely on peerDependenciesMeta.optional, and I need to document for my users that they need to define their schemas using typebox, but pass them to an adapter from an external library (either typeschema or a small adapter from trpc-cli).

Shipping an adapter as part of typebox is how effect solve the same problem: https://trpc.io/docs/server/validators#with-effect

Also some of the links towards the end seem to be out of date now:

1. https://github.com/sinclairzx81/type-adapters 404s
2. https://github.com/sinclairzx81/typebox-adapter has become https://github.com/sinclairzx81/typemap but it was recently marked archived without a note - is it being replaced by something?

---

If you'd be open to adding an adapter, I'd happily contribute a pull request - let me know!

**Reply by sinclairzx81** [reply]
@mmkal Hi, sorry for the delay. Just noticed the new comment here.

---

> A few questions for you, if you're up for reopening this discussion:

Happy to discuss, but almost tempted to move the topic to new a thread as things have evolved somewhat since Jan 2025 (i.e. Standard JSON Schema), and would be good to keep information ordered chronologically (things can clutter in nested threads)

Will draft some notes below and see how things go.

---

> The main one being: would you consider shipping the adapter you linked to here as part of typebox?

Probably not, but happy to discuss anyway. However, before considering the adapter, I would actually be more open to a broader discussion as to why Standard Schema is considered a better integration standard than native Json Schema, ideally using the following code as a reference / talking point ...

```typescript
// Before Standard Schema

server.post('/', {
  body: { type: 'string' }       // Ok:      Server accepts Json Schema. Server uses internal
}, input => { ... })             //          Json Schema validator (i.e. Ajv) which ensures  
                                 //          spec compliance, cross system data-integrity,   
                                 //          vetted security and performance at I/O boundaries.

// After Standard Schema

server.post('/', {
  body: { type: 'string' }       // Error:   'Json Schema' is not 'Standard Json Schema'. Either
}, input => { ... })             //          implement the '~standard' property, or just use
                                 //          Zod, Valibot or ArkType ... >_>
                
```

Given the above, I am fairly skeptical as to whether Standard Schema even has a future, but determining its long term viability would be the primary factor determining whether TB ends up providing an adapter for it. Here are some additional considerations / talking points.

---

### Is Standard Schema Redundant?

If it is possible to infer a Json Schema in the following way ...

Ref: [Basic Json Schema](https://tsplay.dev/wXBMLm) | [Advanced Mcp Protocol Inference Mapping](https://tsplay.dev/wXB7Om)

```typescript
type Vector = Static<{                 // type Vector = {
  type: 'object',                      //   x: number;
  required: ['x', 'y', 'z'],           //   y: number;
  properties: {                        //   z: number;
    x: { type: 'number' },             // }
    y: { type: 'number' },
    z: { type: 'number' }
  }
}>
```

... and it is also possible to infer Zod, Valibot and Arktype via Json Schema transformation

Ref: [Zod](https://tsplay.dev/mLenVW) | [Valibot](https://tsplay.dev/wOYydm) |  [ArkType](https://tsplay.dev/w1reXN)

```typescript
const T = toJsonSchema(z.object({     // const T: {
  x: z.number(),                      //   $schema: "https://json-schema.org/draft/2020-12/schema",
  y: z.number(),                      //   type: "object",
  z: z.number(),                      //   required: ["x", "y", "z"],
}))                                   //   properties: {
                                      //     x: { type: "number" },
                                      //     y: { type: "number" },
                                      //     z: { type: "number" }
                                      //   }
                                      // } = ...

T.type                                // "object"
T.properties                          // { x: ..., y: ..., z: ... }
T.required                            // ['x', 'y', 'z']

type T = Static<typeof T>             // type T = {
                                      //   x: number;
                                      //   y: number;
                                      //   z: number;
                                      // }
```

Then why would a framework choose Standard Schema over Json Schema as the primary interface?

Ref: [The Redundancy of Standard Schema](https://tsplay.dev/WKjYDm)

```typescript
server.post('/', {
  body: z.string().toJsonSchema() // Why do we need Standard Schema when a framework could just
}, input => { ... })              // accept Json Schema and get all the benefits being aligned
                                  // to an international specification + type inference?
```

There are other standardization efforts underway that are worth consideration also ...

Ref: [Proposal Json Schema Standardization under Ecma International](https://github.com/orgs/Json-schema-org/discussions/938)

---

I am open to discussion on all of this, but these days I am usually more compelled to write tooling that allows me to work around Standard Schema. But if you do have a good argument or case to make for it, I am happy to give consideration to it.

---

> Also some of the links towards the end seem to be out of date now:

The project below is the direct successor to TypeMap and TypeAdapters (these projects were retired when TB 1.x landed)

Ref: https://github.com/sinclairzx81/typedriver

For any systems already on Standard Schema, TypeDriver would be TypeBox's direct response to it.

---

Hope to hear back.
S

**Reply by mmkal** [reply]
I think the above is missing the value of standard schema somewhat. It's not competitive with JSON schema (in fact, it is explicitly complementary: [Standard JSON schema](https://standardschema.dev/json-schema) now exists).

JSON schema is a serialisation standard, standard-schema is a validation (and/or transformation) implementation standard.

A simple example. I have a small library called "schematch". It relies on standard schema to provide pattern-matching:

```ts
import {match} from 'schematch'
import {z} from 'zod'

const output = match(input)
  .case(z.string(), s => `hello ${s.slice(1, 3)}`)
  .case(z.array(z.number()), arr => `got ${arr.length} numbers`)
  .case(z.object({msg: z.string()}), obj => obj.msg)
  .default(() => 'unexpected')
```

I can't "replace" standard-schema with json-schema there, because I don't ship a json-schema validator in the library. In fact, I have no dependencies whatsoever. Not zod, not valibot, not arktype. But I support all of them easily because of standard-schema. Those libraries agreed to implement standard-schema, and schematch just says "ok, I will defer the implementation of validation to you".

So replacing with json-schema would turn my library into a completely different one: it'd become a library with big dependencies along with their own supply-chain risk, and its own opinion about which json-schema validation library is best/most correct/most performant. I don't want to impose that on my users.

All of zod, valibot and arktype export to JSON schema via standard-json-schema. But of course, typebox is arguably the best validation library because it sticks so closely to the standards of json-schema and typescript. It has done the hard part! The *easy* part is shipping this very small adapter that makes it play nicely with libraries like `schematch` and `trpc-cli` without them having to impose typebox as an opinion on end-users.

---

So basically, I think your argument applies well to *applications*, which can just make a decision between validation libraries (typebox vs ajv vs zod vs valibot etc.). But it does not apply to *libraries* like trpc, orpc, trpc-cli, schematch etc. - all of which I believe are correct in wanting to remain dependency-free. I would really like to know what you'd suggest for those kinds of libraries that want to have end-users pass them an actual runnable validation *function* - not just a serialisable representation of the schema to be validated against.

A few more examples just to show - I do not think you could just "swap in" json-schema, even though these libraries do rely on being able to convert a zod schema (or valibot or arktype) into json-schema for various introspection/serialisation features

trpc:

```ts
import {initTRPC} from '@trpc/server'
import {z} from 'zod'

const t = initTRPC.create()

const router = t.router({
  sayHello: t
    .input(
      z.object({ name: z.string() })
    )
    .query(({ input }) => `hi ${input.name}`)
})
```

orpc:

```ts
import {os} from '@orpc/server'
import {z} from 'zod'

const router = {
  sayHello: os
    .input(
      z.object({ name: z.string() })
    )
    .handler(({ input }) => `hi ${input.name}`)
}
```

trpc-cli


```ts
import {initTRPC} from '@trpc/server'
import {z} from 'zod'
import {createCli} from 'trpc-cli'

const t = initTRPC.create()

const router = t.router({
  sayHello: t
    .input(
      z.object({ name: z.string() })
    )
    .query(({ input }) => `hi ${input.name}`)
})

createCli({router}).run()
```

**Comment by marcomuser** [comment]
Thank you very much for providing such a detailed response. What you're saying makes sense to me! I wasn't aware of the implications for json schema tooling compatibility.

Since you're not planning to implement it, would it make sense to convert this issue to a discussion (instead of closing it) so that your arguments are somewhat documented? I have a feeling that this will not be the last time that someone will wonder how simple it would be for typebox to implement the interface.

**Comment by sinclairzx81** [comment]
@marcomuser Hi, sorry for the delay in response (busy start to the year)

> Since you're not planning to implement it, would it make sense to convert this issue to a discussion (instead of closing it) so that your arguments are somewhat documented?

Ah, will keep this issue open for a while. Keeping it as an issue helps to make it a bit more visible as well as reminding me to keep an eye on the updates to the spec.

As mentioned though, bar substantial changes to the specification to split validation from schematics, an adapter to Standard Schema is going to be the way to go. The current plan is to continue to revise the [reference implementation](https://github.com/sinclairzx81/typebox/tree/master/example/standard) based on updated spec info, but I may ultimately end up publishing the formal implementation to the following project.

https://github.com/sinclairzx81/typemap

```typescript
import { StandardSchema } from '@sinclair/typemap'

const T = StandardSchema(Type.String())
```

Will keep you posted on any changes / updates.
Cheers
S

**Comment by sinclairzx81** [comment]
@marcomuser Hi, quick update

I have started work on a new project called TypeMap. This project offers advanced translation between libraries and supports Standard Schema by way of projects `Compile` function. This function makes supported libraries compatible with Standard Schema and also offers accelerated type checking by allowing remote libraries to run on TypeBox infrastructure.

### Project

[TypeMap Project](https://github.com/sinclairzx81/typemap)


### Example

[Reference](https://www.typescriptlang.org/play/?moduleResolution=99&module=199#code/JYWwDg9gTgLgBAbzgZQJ4DsYEMAeAaOAFVTAFMAhCfOAYQnGABtS4BfOAMynrgHIABAM7B0AY0ZZgUAPQwSpEFjC8AUCtCRYcAF6duIPtogATVStER0g+AC04AXh0A6CACMAVqVEwAFEjgBgXDS0nAWVrYAXM42JgDyHl4wADwIKgE40dpO6ACuIK6kUD4AlARBFQEhFZkxJgBy+YVQANzpcKhZOU1FpeWVQdVBnXXGjQVFbQHaXXkTxSUDlUOBM6PjzW2si0u7S9WsBE7HAHxq5pbWKA4oGNg4PjY7A9XhV8jRAERItXPNBCM-kUCGsgVA2J9zm94IQbsQyJQHshnhVXpcYdFCAlPN5Uu09gTUaEMpiNpN8YTKSsRoQya0KZSCSs1rSevTGYyDmc1NDaDc6AxmD5CCUANq8AB+1iw6GMWCgpgAuk4AG5YRjAOUwUg+XjHJy8RZoiK0aLIbCy+XGZCiAAWCiwADUAIxOABKpEEuUYKTSHP9AaqxLgvzZU0DEcJ1OiYPDkfjL2DoLDDITaa5QA)

![Image](https://github.com/user-attachments/assets/d8c32a6e-7052-4f7a-8803-65fc5716ace8)

**Comment by sinclairzx81** [comment]
@marcomuser Hiya,

I might convert this issue into a discussion as TypeMap has been formally published as well as added to the Standard Schema [supported library](https://standardschema.dev/) list. TypeMap will be the primary way of integrating TypeBox with Standard Schema moving forward, as well as serve as a vehicle for innovation in the type inference / mapping space.

### Project

[TypeMap Project](https://github.com/sinclairzx81/typemap)

---

Happy to keep the conversation going, feel free to reach out if you have anything thoughts or suggestions on the design.
Cheers
S

**Reply by marcomuser** [reply]
Perfect! This looks awesome. Thank you for your effort!!

**Comment by martiinii** [comment]
Hey! I'm trying to switch from TypeBox to ArkType (ctx: Elysia.js, trying to keep my schemas in separate package, monorepo setup). The issue with Elysia is, it relies heavily on TypeBox. I've been trying to convert ArkType into TypeBox, however `Syntax` always returns `never`. ArkType already support standard schema, what could be the issue?
```ts
const myType = type({
  x: "number"
})

const S = Syntax(myType) // "never"
const T = TypeBox(S)
```

**Reply by sinclairzx81** [reply]
@Martiinii Hi,

> ArkType already support standard schema, what could be the issue?

TypeMap only supports Zod and Valibot as their runtime structures were mappable. An attempt was made to support ArkType but wasn't feasible as it's runtime representations were opaque (and thus not mappable). Refer to the link below for more information.

https://github.com/standard-schema/standard-schema/pull/35#issuecomment-2614523789

Keep in mind, TypeMap does not map from Standard Schema because Standard Schema is NOT a schema, it is a TS interface with a `validate()` function that contains no runtime information (and thus is non-portable/mappable). Unfortunately the naming of Standard Schema is going to mislead a lot of developers. 

---

> The issue with Elysia is, it relies heavily on TypeBox

I don't consider this to be an issue. TypeBox is best in class, however there is work being done to enable Elysia (and other frameworks) to support a wider range of libraries (including ArkType) by implementing extensions to TypeBox's validation infrastructure. This work should be ready later 2025.

**Reply by martiinii** [reply]
Thank you for detailed answer! Yeah, the standard schema naming is confusing. I don't mean to hate on Elysia or TypeBox - both are amazing, but I thought Elysia already supports standard schema - guess I'll be patient
